AX: VoiceOver focus pops out of the text field or gets stuck in PDFs https://bugs.webkit.org/show_bug.cgi?id=289190 rdar://146323940

utkuuckun opened the pull request:
#### d8e4df8c3919e26415b9ffb33f0e9bed0fc2d8cc
<pre>
AX: VoiceOver focus pops out of the text field or gets stuck in PDFs <a href="https://bugs.webkit.org/show_bug.cgi?id=289190">https://bugs.webkit.org/show_bug.cgi?id=289190</a> <a href="https://rdar.apple.com/146323940">rdar://146323940</a>

Reviewed by Tyler Wilcock.

-We want to build isolated tree for the main frame plugins so accessibility objects can be retrieved by assistive technologies. But we don&apos;t want to set isolated tree as root because we want PDF plugin to handle requests.
-VoiceOver expects to receive PDF annotation element as focusedUIElement instead of the WebKit text box. Therefore, when accessibilityFocusedUIElement is queried we want to return the AXAssociatedPluginParent if one exists.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::buildAccessibilityTreeIfNeeded):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:]):
(-[WKAccessibilityWebPageObjectBase setIsolatedTreeRoot:]):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::accessibilityFocusedUIElement):

Canonical link: <a href="https://commits.webkit.org/291756@main">https://commits.webkit.org/291756@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21ee123401ec1663dbe45724290c53e23a668f20

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93841 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13502 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3156 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98847 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44367 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95891 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13753 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21857 "Built successfully") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71615 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-layout-tests-in-stress-mode; 76 flakes 119 failures; Uploaded test results; 22 flakes 69 failures; Compiled WebKit (warnings); 1 flakes 68 failures; Running analyze-layout-tests-results") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28983 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96843 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10217 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84785 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51951 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9898 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2420 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43704 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80163 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2493 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100883 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20893 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15239 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80632 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21145 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80739 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79978 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19924 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24526 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1883 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14050 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20877 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26055 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20564 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24024 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22305 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -109,8 +109,13 @@ - (id)accessibilityRootObjectWrapper:(WebCore::LocalFrame*)frame
+#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)
+            if (auto cache = protectedSelf.get().axObjectCache)
+                cache->buildAccessibilityTreeIfNeeded();
+#endif
```
Could we add a comment explaining why we do this? Something to the effect of:

"Even though we want to serve the PDF plugin tree for main-frame plugins, we still need to make sure the isolated tree is built, so that when text annotations are created on-the-fly as users focus on text fields, isolated objects are able to be attached to those text annotation object wrappers. If they aren't, we never have a backing object to serve any requests from."

twilco replied:
Also could we use `CheckedPtr cache ` instead of `auto cache`?

Commit:
AX: VoiceOver focus pops out of the text field or gets stuck in PDFs https://bugs.webkit.org/show_bug.cgi?id=289190 rdar://146323940

Reviewed by Tyler Wilcock.

-We want to build isolated tree for the main frame plugins so accessibility objects can be retrieved by assistive technologies. But we don't want to set isolated tree as root because we want PDF plugin to handle requests.
-VoiceOver expects to receive PDF annotation element as focusedUIElement instead of the WebKit text box. Therefore, when accessibilityFocusedUIElement is queried we want to return the AXAssociatedPluginParent if one exists.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::buildAccessibilityTreeIfNeeded):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:]):
(-[WKAccessibilityWebPageObjectBase setIsolatedTreeRoot:]):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::accessibilityFocusedUIElement):

Canonical link: https://commits.webkit.org/291756@main
